### PR TITLE
test(conformance): tls conformance test enabled

### LINKED ISSUE #3678: extended conformance - TLSRoute
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Problem Statement

Support the `TLSRoute` features in conformance tests (extended)

### Proposed Solution

_No response_

### Additional information

_No response_

### Acceptance Criteria

- [ ] the test `TLSRouteSimpleSameNamespace` is successfully ran

### PR DESCRIPTION
**Pull request by mlavacca**
**What this PR does / why we need it**:

**Which issue this PR fixes**:

Fixes #3678

**Special notes for your reviewer**:

We had `TLSRoute` integration tests with both Passthrough and terminate, but KIC [currently supports Passthrough only](https://github.com/Kong/kubernetes-ingress-controller/blob/caf7f1a25a8be9e9a2dba30aab04bdb33ab629d8/internal/controllers/gateway/route_utils.go#L384-L389) due to an [upstream docs inconsistency](https://github.com/kubernetes-sigs/gateway-api/issues/1474). 
The integration test with terminate was actually using a gateway with passthrough (we were not really testing the terminate mode). For this reason, the integration test with terminate has been removed, and the one with Passthrough has been massively improved.

**PR Readiness Checklist**:

Complete these before marking the PR as `ready to review`:

- [x] the `CHANGELOG.md` release notes have been updated to reflect any significant (and particularly user-facing) changes introduced by this PR

### CODE REVIEW

**Comment by pmalek** [comment]
@mlavacca Do we still need this 🤔 ?

**Comment by mlavacca** [comment]
> @mlavacca Do we still need this 🤔 ?

Yes, we do. I didn't manage to find time for it, I'll get back to it quite soon

**Comment by rainest** [comment]
@mlavacca it looks like this got sidelined again. I was going to try and pick it up but wasn't sure what was remaining or exactly what was broken before. Can you explain what the change was trying to do?

It looks like we can't actually satisfy the conformance test in KIC alone. It requires serving the TLSRoute on port 443, whereas the test instance has hardcoded HTTPS listens on 443 and stream TLS listens on 8899.  We get the validation test cert on 443 (not entirely sure where the config setting that is, but we do) and the passthrough cert (what this test wants) on 8899:

```
17:07:59-0700 esenin $ openssl s_client -connect 172.18.0.102:8899 -servername abc.example.com -showcerts < /dev/null | openssl x509 -text | grep -A1 Alternative
depth=0 O = Acme Co, CN = default
verify error:num=18:self-signed certificate
verify return:1
depth=0 O = Acme Co, CN = default
verify return:1
DONE
            X509v3 Subject Alternative Name: 
                DNS:abc.example.com

17:09:11-0700 esenin $ openssl s_client -connect 172.18.0.102:443 -servername abc.example.com -showcerts < /dev/null | openssl x509 -text | grep -A1 Alternative 
depth=0 O = Acme Co, CN = default
verify error:num=18:self-signed certificate
verify return:1
depth=0 O = Acme Co, CN = default
verify return:1
DONE
            X509v3 Subject Alternative Name: 
                DNS:*, DNS:*.org
```

I'm not positive we violate conformance there (the spec probably _does_ want support for both HTTPS and TLS on the same port), but I am fairly certain that the gateway can only serve stream routes on stream listens.

We could satisfy the test (since it only has the one listen and doesn't need both HTTPS and TLS on the same port) with the operator dynamically provisioning kong.conf listen configuration, but we can't satisfy it with the static config we use in the KIC tests.

**Comment by mlavacca** [comment]
> @mlavacca it looks like this got sidelined again. I was going to try and pick it up but wasn't sure what was remaining or exactly what was broken before. Can you explain what the change was trying to do?
> 
> It looks like we can't actually satisfy the conformance test in KIC alone. It requires serving the TLSRoute on port 443, whereas the test instance has hardcoded HTTPS listens on 443 and stream TLS listens on 8899. We get the validation test cert on 443 (not entirely sure where the config setting that is, but we do) and the passthrough cert (what this test wants) on 8899:
> 
> ```
> 17:07:59-0700 esenin $ openssl s_client -connect 172.18.0.102:8899 -servername abc.example.com -showcerts < /dev/null | openssl x509 -text | grep -A1 Alternative
> depth=0 O = Acme Co, CN = default
> verify error:num=18:self-signed certificate
> verify return:1
> depth=0 O = Acme Co, CN = default
> verify return:1
> DONE
>             X509v3 Subject Alternative Name: 
>                 DNS:abc.example.com
> 
> 17:09:11-0700 esenin $ openssl s_client -connect 172.18.0.102:443 -servername abc.example.com -showcerts < /dev/null | openssl x509 -text | grep -A1 Alternative 
> depth=0 O = Acme Co, CN = default
> verify error:num=18:self-signed certificate
> verify return:1
> depth=0 O = Acme Co, CN = default
> verify return:1
> DONE
>             X509v3 Subject Alternative Name: 
>                 DNS:*, DNS:*.org
> ```
> 
> I'm not positive we violate conformance there (the spec probably _does_ want support for both HTTPS and TLS on the same port), but I am fairly certain that the gateway can only serve stream routes on stream listens.
> 
> We could satisfy the test (since it only has the one listen and doesn't need both HTTPS and TLS on the same port) with the operator dynamically provisioning kong.conf listen configuration, but we can't satisfy it with the static config we use in the KIC tests. We should probably still add whatever improvement was proposed here, but it won't actually help with that test.

Yes, this was my figuring as well before putting this one aside. I'd like to revamp it and see if what is introduced here is still needed and figure out how we can dynamically configure Kong ports to make this test pass. This is in the 2.11 milestone, hence has quite a high priority now.

File: CHANGELOG.md [hunk 1]
```diff
@@ -76,7 +76,10 @@ Adding a new version? You'll need three changes:
 ### Added
 
 - **WIP** Introduce `KongConsumerGroup` CRD (supported by Kong Enterprise only)
-  [#4325](https://github.com/Kong/kubernetes-ingress-controller/pull/4325), [#4387](https://github.com/Kong/kubernetes-ingress-controller/pull/4387), [#4419](https://github.com/Kong/kubernetes-ingress-controller/pull/4419), [4437](https://github.com/Kong/kubernetes-ingress-controller/pull/4437)
+  [#4325](https://github.com/Kong/kubernetes-ingress-controller/pull/4325)
+  [#4387](https://github.com/Kong/kubernetes-ingress-controller/pull/4387)
+  [#4419](https://github.com/Kong/kubernetes-ingress-controller/pull/4419)
+  [4437](https://github.com/Kong/kubernetes-ingress-controller/pull/4437)
```
Comment (programmer04) [inline CHANGELOG.md:82]:
```suggestion
  [#4437](https://github.com/Kong/kubernetes-ingress-controller/pull/4437)
```

**Review by programmer04 — COMMENTED** [review]
LGTM, but someone more seasoned should take a look too

**Review by rainest — COMMENTED** [review]
Should we actually need a separate TLS publish service CLI argument and expect a separate service for it?

The reason we have that for UDP is that dual-transport LoadBalancers aren't necessarily available until we drop compatibility for Kubernetes <=1.25 (see MixedProtocolLBService in https://kubernetes.io/docs/reference/command-line-tools-reference/feature-gates/#feature-gates-for-graduated-or-deprecated-features), but TLS should generally be fine on the main publish service as it's still TCP at its base.

There may be reason to separate that at the cloud provider level (e.g. AFAIK AWS NLBs can operate in either terminate or passthrough), but I'm not sure that affects us (if your LB does termination, the connection Kong sees is plain TCP) and the PR doesn't mention it as being a concern.

While the conformance test [conflicts with our standard configuration](https://github.com/Kong/kubernetes-ingress-controller/pull/3797#issuecomment-1635079225), I don't think we'd want to introduce a separate Service config just for that. The Service should be able to support both stream TLS and HTTPS so long as they're on different ports, but I'm not clear on whether compliance wants support for them on the same port (in which case we're out of luck--the gateway doesn't support same-port HTTP(S) and stream listens) or just has issues because we can't dynamically configure a stream listener on 443 (but we can create a separate Service that directs 443 to some other stream port).

**Comment by randmonkey** [comment]
> Should we actually need a separate TLS publish service CLI argument and expect a separate service for it?
> 
> The reason we have that for UDP is that dual-transport LoadBalancers aren't necessarily available until we drop compatibility for Kubernetes <=1.25 (see MixedProtocolLBService in https://kubernetes.io/docs/reference/command-line-tools-reference/feature-gates/#feature-gates-for-graduated-or-deprecated-features), but TLS should generally be fine on the main publish service as it's still TCP at its base.
> 
> There may be reason to separate that at the cloud provider level (e.g. AFAIK AWS NLBs can operate in either terminate or passthrough), but I'm not sure that affects us (if your LB does termination, the connection Kong sees is plain TCP) and the PR doesn't mention it as being a concern.
> 
> While the conformance test [conflicts with our standard configuration](https://github.com/Kong/kubernetes-ingress-controller/pull/3797#issuecomment-1635079225), I don't think we'd want to introduce a separate Service config just for that. The Service should be able to support both stream TLS and HTTPS so long as they're on different ports, but I'm not clear on whether compliance wants support for them on the same port (in which case we're out of luck--the gateway doesn't support same-port HTTP(S) and stream listens) or just has issues because we can't dynamically configure a stream listener on 443 (but we can create a separate Service that directs 443 to some other stream port).

I think we do not need a separate LB service for TLS other than TCP. IMO For TLS mode we can simply configure them as `passthrough` to pass the original TCP packets to Kong.

**Comment by mlavacca** [comment]
> While the conformance test https://github.com/Kong/kubernetes-ingress-controller/pull/3797#issuecomment-1635079225, I don't think we'd want to introduce a separate Service config just for that. The Service should be able to support both stream TLS and HTTPS so long as they're on different ports, but I'm not clear on whether compliance wants support for them on the same port (in which case we're out of luck--the gateway doesn't support same-port HTTP(S) and stream listens) or just has issues because we can't dynamically configure a stream listener on 443 (but we can create a separate Service that directs 443 to some other stream port).

This exact issue is what made me introduce such a new service. The Conformance tests use a gateway with TLS with passthrough on port 443 (the same port used by other HTTPS tests), and I found no way to configure both HTTPS and TLS stream on the same port. This is why I created such a service that uses port 443 and redirects traffic to port 8899 of the Gateway.

**Comment by mlavacca** [comment]
> I think we do not need a separate LB service for TLS other than TCP. IMO For TLS mode we can simply configure them as passthrough to pass the original TCP packets to Kong.

Unfortunately, this does not look possible in this context (see the comment above).

**Comment by randmonkey** [comment]
So the new service for TLS is required, since conformance test uses the fixed port 443 (the same as HTTPS port). I agree that we need a new service under this condition.

**Comment by rainest** [comment]
tl;dr we shouldn't do this just because the test is incompatible with our preset configuration, but there are other (untested) parts of the spec that would require it, so we probably should move forward with the separate publish Service.

For this test alone, I'd argue that we not add a separate publish service, because this failure only happens because of the port collision, where we configure an HTTPS listen on that port already, and cannot mix NGINX `http` and `stream` handling. If either:
- We had dynamic listen configuration, where the operator spawns a Kong instance with the exact listens requested in a Gateway only.
- https://github.com/kubernetes-sigs/gateway-api/blob/main/conformance/tests/tlsroute-simple-same-namespace.yaml requested a port other than 443, and we configured a stream listen for that port in the KTF configuration.

the test would otherwise pass.

However, that test doesn't fully cover the spec. It only requests a TLS listen in isolation, whereas [the spec](https://gateway-api.sigs.k8s.io/references/spec/#gateway.networking.k8s.io/v1beta1.GatewaySpec) has

> An implementation MAY group Listeners by Port and then collapse each group of Listeners into a single Listener if the implementation determines that the Listeners in the group are “compatible”
> For example, an implementation might consider Listeners to be compatible with each other if all of the following conditions are met:
> - Either each Listener within the group specifies the “HTTP” Protocol or each Listener within the group specifies either the “HTTPS” or “TLS” Protocol.
> - ... (remaining conditions aren't relevant here)

There is no conformance test that spawns both a TLS and HTTPS Listener on the same port, but the spec indicates this is expected configuration. Collapsing these is only a `MAY`, however, which I'm interpreting to mean that separate IPs (the end effect of this PR's change) are compliant.

Documentation for this feature may be a bit confusing, since you'd only need the separate Service if you do indeed want to share ports between HTTPS and TLS. If not, you can just use the main Service for any TCP-based traffic.

I am curious [if there are other implementations doing the same](https://kubernetes.slack.com/archives/CR0H13KGA/p1690999473709589), but my intuition is that some are.

File: test/conformance/gateway_conformance_test.go [hunk 2]
```diff
@@ -39,7 +42,7 @@ var skippedTestsForExpressionRoutes = []string{
 	tests.HTTPRouteRedirectPortAndScheme.ShortName,
 	// https://github.com/Kong/kubernetes-ingress-controller/issues/3680
 	tests.GatewayClassObservedGenerationBump.ShortName,
-	// https://github.com/Kong/kubernetes-ingress-controller/issues/3678
+	// https://github.com/Kong/kubernetes-ingress-controller/issues/4312
```
Comment (rainest) [inline test/conformance/gateway_conformance_test.go:45]: What's the remaining reason for omitting this test until we have expression route support? Can we add a clarifying comment as to why?

I thought that was only necessary when we needed to manage rule or cross-Route priority order, but https://github.com/kubernetes-sigs/gateway-api/blob/main/conformance/tests/tlsroute-simple-same-namespace.yaml only has the one single-rule TLSRoute.

**Review by rainest — CHANGES_REQUESTED** [review]
For posterity, what were the improvements to the integration test? The diff is, unfortunately, quite confusing with the terminate test getting removed.

The new logic in `determineServiceForGateway()` does account for the TLS listener being optional (which we do want), but we can only test one variant in integration, so I'd want an envtest to check both the TLS publish service present and absent cases if possible.

---

Edit: attempted to write [an envtest](https://gist.github.com/rainest/819f6f759ad1e2946757e0d7c42aa8e0) for this, but it looks like it's not possible. Apparently the envtest skeleton environment will strip status information from Services if you try and pre-populate it, so Services will never have any addresses to insert into the Gateway status.

Not sure why that happens. I also attempted adding a `client.Update(ctx, &svc)` after, and no luck there either. I would expect those to work, given that we use the same client functions in the controllers, and those _do_ manage to apply status updates.

Writing those did raise another issue, that addresses are per-Gateway rather than per-Listener. If you use a TLS publish service and create both TLS and other Listeners on the same Gateway, the the listed Gateway addresses will only serve a subset of the attached routes. We do the same thing for UDP, however, and apparently didn't discuss it [when adding that](https://github.com/Kong/kubernetes-ingress-controller/pull/3325).

AFAICT we can only do address status properly if you split Listeners that require their own Service onto a separate Gateway. This doesn't really exacerbate that problem though, so I don't think it should block this, annoying as it is.

File: test/conformance/gateway_conformance_test.go (hunk 2 above)
Comment (mlavacca) [reply test/conformance/gateway_conformance_test.go:45]: The remaining reason for this test is that expression routes do not support L4 routes yet. #4312 is about it.

**Comment by mlavacca** [comment]
> For posterity, what were the improvements to the integration test? The diff is, unfortunately, quite confusing with the terminate test getting removed.

The integration test has been basically fixed rather than improved. The following were the problems:
- we were using a gateway with TLS mode passthrough, but actually, we were terminating the connection in Kong.
- we had tests with terminate, but TLS mode termination is not currently supported for `TLSRoutes` because of https://github.com/kubernetes-sigs/gateway-api/issues/1474

For these reason:
- the terminate test has been deleted
- the passthrough test has been improved with all the additional checks we were doing in the terminate test
- the `referenceGrant` test has been set to use passthrough as well.

**Comment by mlavacca** [comment]
> Edit: attempted to write [an envtest](https://gist.github.com/rainest/819f6f759ad1e2946757e0d7c42aa8e0) for this, but it looks like it's not possible. Apparently the envtest skeleton environment will strip status information from Services if you try and pre-populate it, so Services will never have any addresses to insert into the Gateway status.

Not sure why that happens. I also attempted adding a client.Update(ctx, &svc) after, and no luck there either. I would expect those to work, given that we use the same client functions in the controllers, and those do manage to apply status updates.

Writing those did raise another issue, that addresses are per-Gateway rather than per-Listener. If you use a TLS publish service and create both TLS and other Listeners on the same Gateway, the the listed Gateway addresses will only serve a subset of the attached routes. We do the same thing for UDP, however, and apparently didn't discuss it https://github.com/Kong/kubernetes-ingress-controller/pull/3325.
>
> AFAICT we can only do address status properly if you split Listeners that require their own Service onto a separate Gateway. This doesn't really exacerbate that problem though, so I don't think it should block this, annoying as it is.

To not lose track of this problem, should we create an issue to address it?

File: test/conformance/gateway_conformance_test.go (hunk 2 above)
Comment (rainest) [reply test/conformance/gateway_conformance_test.go:45]: :facepalm: I only just now noticed we had a separate `skippedTestsForExpressionRoutes` array and thought there was only the one.

**Review by rainest — DISMISSED** [review]

**Comment by rainest** [comment]
Per upstream:

> If that's not possible for an implementation, then the Gateway should not be Accepted, since the implementation can't satisfy the request.

So we should probably not have this, but given the existence of the equivalent UDP argument (albeit for different reasons) and lack of formal language/tests in the spec, the actual correct behavior is kinda in limbo :|

**Comment by mlavacca** [comment]
@rainest you approved but then re-asked yourself for the review. WDYT, can we merge this?

**Review by shaneutt — DISMISSED** [review]
I did some brainstorming on this one with @mlavacca today.

Our dilemma is that we have a static-mode Gateway API implementation in KIC, which means we are basically at odds with the dynamism the tests are expecting from Gateway `Listeners` (flatly, we **can't** be dynamic at all we can only be preemptive with Helm chart configuration). There are a few ways we could resolve this, including trying to ask for changes in upstream to either allow the port for the test to be changed dynamically, OR ask for the static port to be changed but these come across as "asking upstream to put in hacks for Kong". I think we should leave such a change upstream off the table.

There are a few different alternatives we thought up to solve this problem, but they all end up being "hack around our static mode limitations", and ultimately they don't deliver on "allowing users to use port 443 for TLSRoute". @mlavacca's solution here of following `UDPRoute`'s example for `TLSRoute` strikes me as pragmatic, reasonably simple and rooted in precedent (so therefore we're being somewhat consistent) while also avoiding "hacks in upstream". This will be consistent with how we've enabled something like this in the past, and it will actually _enable_ it rather than working around it by changing the port somewhere.

So all things considered, I think this is a reasonable solution.

**Comment by mlavacca** [comment]
Closing as we'll implement TLS conformance in [KGO instead](https://github.com/Kong/gateway-operator/issues/65), where Dynamic listener configuration is not an issue as it is in KIC.
